### PR TITLE
[fx] split Node into Node/Proxy

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #43083 [fx] add type annotations
* #43082 [fx] enabling typechecking of fx files
* **#42991 [fx] split Node into Node/Proxy**

Have Node both be a record of the operator in the graph, and the
way we _build_ the graph made it difficult to keep the IR datastructure
separate from the proxying logic in the build.

Among other issues this means that typos when using nodes would add
things to the graph:
```
    for node in graph.nodes:
        node.grph # does not error, returns an node.Attribute object!
```

This separates the builder into a Proxy object. Graph/Node no longer
need to understand `delegate` objects since they are now just pure IR.
This separates the `symbolic_trace` (proxy.py/symbolic_trace.py) from
the IR (node.py, graph.py).

This also allows us to add `create_arg` to the delegate object,
allowing the customization of how aggregate arguments are handled
when converting to a graph.

Differential Revision: [D23099786](https://our.internmc.facebook.com/intern/diff/D23099786)

### CODE REVIEW

File: torch/fx/symbolic_trace.py [hunk 1]
```diff
@@ -53,13 +65,51 @@ class DefaultDelegate:
...
+
+        raise NotImplementedError(f"argument of type: {type(a)}")
+
+
+class DefaultDelegate(DelegateBase):
+    def __init__(self, root: torch.nn.Module, graph: Graph):
+        super().__init__(graph)
+        self.root = root
+
+    def create_arg(self, a):
```
Comment (jamesr66a) [inline torch/fx/symbolic_trace.py:101]: nit: Trailing whitespace on this line

File: torch/fx/symbolic_trace.py [hunk 2]
```diff
@@ -53,13 +65,51 @@ class DefaultDelegate:
...
+    def __init__(self, root: torch.nn.Module, graph: Graph):
+        super().__init__(graph)
+        self.root = root
+
+    def create_arg(self, a):        
+        if isinstance(a, torch.nn.Parameter):
+            for n, p in self.root.named_parameters():
+                if a is p:
+                    return self.graph.get_param(n)
+            raise NameError('parameter is not a member of this module')
```
Comment (jamesr66a) [inline torch/fx/symbolic_trace.py:106]: What's going on here? Isn't this covered in the base implementation?

**Review by jamesr66a — APPROVED** [review]
LGTM, only small comments

File: torch/fx/symbolic_trace.py (hunk 2 above)
Comment (zdevito) [reply torch/fx/symbolic_trace.py:106]: Yeah, I refactored wrong in this revision. The fixed version is now more clear: the base delegate cannot handle parameters because it doesn't have a `root` module associated with it. It is used primarily when you are just using the proxy object to add to a graph outside of symbolic tracing, e.g. `t = Proxy(node); output = (t *t - t).node`.

Comment (zdevito) [reply torch/fx/symbolic_trace.py:106]: The advantage of the "only Node" approach before was that you didn't have to wrap nodes in proxy objects to add more compute with them. But the problems it introduces out-weigh making it slightly more difficult to do.
